cogs5e/charGen.py

```diff
@@ -8,15 +8,121 @@
 from discord.ext import commands
 
 from cogs5e.models import embeds
-from cogs5e.models.embeds import EmbedWithAuthor
+from cogs5e.models.embeds import EmbedWithAuthor, EmbedWithColor
 from cogs5e.models.errors import InvalidArgument
 from gamedata.compendium import compendium
 from gamedata.lookuputils import available, get_race_choices
+from utils.constants import STAT_ABBREVIATIONS
 from utils.functions import get_selection, search_and_select
 
 log = logging.getLogger(__name__)
 
 
+async def roll_stats(ctx):
+    guild_settings = await ctx.get_server_settings()
+
+    dice = "4d6kh3"
+    sets = 1
+    stats = 6
+    straight = False
+    min_total = None
+    max_total = None
+    over = {}
+    under = {}
+    stat_names = [stat.upper() for stat in STAT_ABBREVIATIONS]
+
+    if guild_settings:
+        dice = guild_settings.randchar_dice
+        sets = guild_settings.randchar_sets
+        stats = guild_settings.randchar_num
+        straight = guild_settings.randchar_straight
+        min_total = guild_settings.randchar_min
+        max_total = guild_settings.randchar_max
+        stat_names = guild_settings.randchar_stat_names or stat_names
+        if guild_settings.randchar_rules:
+            for rule in guild_settings.randchar_rules:
+                if rule.type == "gt":
+                    over[rule.value] = rule.amount
+                if rule.type == "lt":
+                    under[rule.value] = rule.amount
+
+    embed = EmbedWithColor()
+    embed.title = "Generating Random Stats"
+
+    # Generate our rule text
+    rules = []
+    if sets > 1:
+        rules.append(f"Rolling {sets} sets")
+    if straight:
+        rules.append("Assigning stats directly")
+    if stats != 6:
+        rules.append(f"Rolling {stats} per set")
+    if min_total:
+        rules.append(f"Minimum of {min_total}")
+    if max_total:
+        rules.append(f"Maximum of {max_total}")
+    for m, t in over.items():
+        rules.append(f"At least {t} over {m}")
+    for m, t in under.items():
+        rules.append(f"At least {t} under {m}")
+
+    stat_rolls = []
+    # Only attempt 1000 times to achieve the desired stat sets
+    for _ in range(1000):
```

Comment:
I recommend using `utils.dice.PersistentRollContext` with a high limit (maybe max_rolls=5000?) in addition to this for loop; see the implementation of the `!rr` command

Reply:
If you do this you'll probably want to catch the `d20.TooManyRolls` exception too